src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -500,23 +502,26 @@ suggestRangeFinderPrecision maxPages =
 -------------------------------------------------------------------------------}
 
 data SearchResult =
-    NoResult
-  | SinglePage PageNo
+    SinglePage {-# UNPACK #-} !PageNo
   -- | @'Multipage' s e@: the value is larger than a page, starts on page s, and
   -- ends on page e.
-  | MultiPage PageNo PageNo
+  | MultiPage {-# UNPACK #-} !PageNo {-# UNPACK #-} !PageNo
   deriving stock (Show, Eq)
 
 -- | A span of pages, representing an inclusive interval of page numbers.
 data PageSpan = PageSpan {
-    pageSpanStart :: PageNo
-  , pageSpanEnd   :: PageNo
+    pageSpanStart :: {-# UNPACK #-} !PageNo
+  , pageSpanEnd   :: {-# UNPACK #-} !PageNo
   }
+  deriving (Show, Eq)
 
-toPageSpan :: SearchResult -> Maybe PageSpan
-toPageSpan NoResult        = Nothing
-toPageSpan (SinglePage i)  = Just (PageSpan i i)
-toPageSpan (MultiPage i j) = assert (i < j) $ Just (PageSpan i j)
+pageSpanSize :: PageSpan -> NumPages
+pageSpanSize pspan =
+    unPageNo (pageSpanEnd pspan) - unPageNo (pageSpanStart pspan) + 1
+
+toPageSpan :: SearchResult -> PageSpan
```

Comment:
Or perhaps `search` should always return `PageSpan` and we can get rid of `SearchResult`. They are equivalent, but `PageSpan` uses a single constructor, which means it often doesn't need to be allocated at all (constructed product result optimisation).

See `Note [Constructed Product Result]`  :grin:

Reply:
https://github.com/IntersectMBO/lsm-tree/pull/144#discussion_r1551761184